File: agent/crates/osquery-client/src/scheduler.rs
use crate::client::OsqueryClient;
use crate::diff;
use crate::types::{
    ColumnEntry, OsqueryResult, OsqueryResultRow, OsqueryRow, ResultAction, ScheduledQuery,
};
use anyhow::Result;
use chrono::Utc;
use rusqlite::Connection;
use std::path::{Path, PathBuf};
use tokio::sync::mpsc;

pub struct QueryScheduler {
    conn: Connection,
}

impl QueryScheduler {
    pub fn new(db_path: &Path) -> Result<Self> {
        let conn = Connection::open(db_path)?;

        conn.execute(
            "CREATE TABLE IF NOT EXISTS scheduled_queries (
                name TEXT PRIMARY KEY,
                query TEXT NOT NULL,
                interval_secs INTEGER NOT NULL,
                snapshot INTEGER NOT NULL DEFAULT 0,
                updated_at INTEGER NOT NULL
            )",
            [],
        )?;

        let count: i64 = conn.query_row("SELECT COUNT(*) FROM scheduled_queries", [], |row| {
            row.get(0)
        })?;

        if count == 0 {
            let now = chrono::Utc::now().timestamp();
            let defaults = vec![
                (
                    "running_processes",
                    "SELECT pid, name, path, cmdline, uid FROM processes;",
                    30,
                    0,
                ),
                (
                    "listening_ports",
                    "SELECT pid, port, protocol, address FROM listening_ports;",
                    30,
                    0,
                ),
                (
                    "users",
                    "SELECT username, uid, gid, shell, directory FROM users;",
                    300,
                    0,
                ),
            ];
            for (name, query, interval, snapshot) in defaults {
                conn.execute(
                    "INSERT INTO scheduled_queries (name, query, interval_secs, snapshot, updated_at) VALUES (?1, ?2, ?3, ?4, ?5)",
                    rusqlite::params![name, query, interval, snapshot, now],
                )?;
            }
            tracing::info!("Seeded default scheduled queries into empty database");
        }

        Ok(Self { conn })
    }

    pub fn load_queries(&self) -> Result<Vec<ScheduledQuery>> {
        let mut stmt = self
            .conn
            .prepare("SELECT name, query, interval_secs, snapshot FROM scheduled_queries")?;
        let query_iter = stmt.query_map([], |row| {
            let snapshot: i32 = row.get(3)?;
            Ok(ScheduledQuery {
                name: row.get(0)?,
                query: row.get(1)?,
                interval_secs: row.get(2)?,
                snapshot: snapshot != 0,
            })
        })?;

        let mut queries = Vec::new();
        for query in query_iter {
            queries.push(query?);
        }
        Ok(queries)
    }

    pub fn upsert_queries(&mut self, queries: &[ScheduledQuery]) -> Result<()> {
        let tx = self.conn.transaction()?;

        let now = chrono::Utc::now().timestamp();

        for query in queries {
            tx.execute(
                "INSERT INTO scheduled_queries (name, query, interval_secs, snapshot, updated_at) 
                 VALUES (?1, ?2, ?3, ?4, ?5)
                 ON CONFLICT(name) DO UPDATE SET 
                 query=excluded.query, interval_secs=excluded.interval_secs, snapshot=excluded.snapshot, updated_at=excluded.updated_at",
                rusqlite::params![
                    query.name,
                    query.query,
                    query.interval_secs,
                    if query.snapshot { 1 } else { 0 },
                    now,
                ],
            )?;
        }

        tx.commit()?;
        Ok(())
    }

    /// Run the scheduler. Each scheduled query gets its own task with a persistent
    /// OsqueryClient connection that reconnects on error.
    ///
    /// Queries are loaded *before* entering the async context so that the
    /// rusqlite::Connection is never held across an await point.
    pub async fn run(
        self,
        tx: mpsc::Sender<OsqueryResult>,
        socket_path: PathBuf,
        agent_uuid: String,
    ) {
        // Load queries synchronously before dropping self (and its Connection).
        let queries = match self.load_queries() {
            Ok(q) => q,
            Err(e) => {
                tracing::error!("Failed to load scheduled queries from SQLite: {}", e);
                return;
            }
        };
        // Drop self here — Connection is no longer held.
        drop(self);

        if queries.is_empty() {
            tracing::warn!("No scheduled queries found in SQLite — nothing to run.");
            return;
        }

        tracing::info!("Starting {} scheduled query task(s)", queries.len());

        for query in queries {
            let tx = tx.clone();
            let socket_path = socket_path.clone();
            let agent_uuid = agent_uuid.clone();

            tokio::spawn(async move {
                // Connect once per query task; reconnect on error inside the loop.
                let mut client = loop {
                    match OsqueryClient::connect(&socket_path).await {
                        Ok(c) => break c,
                        Err(e) => {
                            tracing::warn!(
                                "[{}] Cannot connect to osquery yet ({}), retrying in 5s...",
                                query.name,
                                e
                            );
                            tokio::time::sleep(std::time::Duration::from_secs(5)).await;
                        }
                    }
                };

                let mut previous_rows: Vec<OsqueryRow> = Vec::new();
                let mut first_run = true;
                let mut interval =
                    tokio::time::interval(std::time::Duration::from_secs(query.interval_secs));

                loop {
                    interval.tick().await;

                    let response = match client.query(&query.query).await {
                        Ok(res) => res,
                        Err(e) => {
                            tracing::warn!("[{}] Query error: {}", query.name, e);
                            // Client will reconnect internally on next call.
                            continue;
                        }
                    };

                    if response.status.code != 0 {
                        tracing::warn!(
                            "[{}] osquery error (code {}): {}",
                            query.name,
                            response.status.code,
                            response.status.message
                        );
                        continue;
                    }

                    let current_rows = response.rows;
                    tracing::debug!("[{}] Got {} rows", query.name, current_rows.len());

                    if query.snapshot {
                        // Snapshot mode: emit all rows every tick.
                        let result = Self::build_result(
                            &query.name,
                            &agent_uuid,
                            current_rows,
                            ResultAction::Snapshot,
                        );
                        if tx.send(result).await.is_err() {
                            tracing::info!(
                                "[{}] Result channel closed, stopping task.",
                                query.name
                            );
                            break;
                        }
                    } else {
                        // Differential mode.
                        if first_run {
                            // First run: emit a full snapshot as the baseline.
                            let result = Self::build_result(
                                &query.name,
                                &agent_uuid,
                                current_rows.clone(),
                                ResultAction::Snapshot,
                            );
                            if tx.send(result).await.is_err() {
                                tracing::info!(
                                    "[{}] Result channel closed, stopping task.",
                                    query.name
                                );
                                break;
                            }
                            first_run = false;
                        } else {
                            let (added, removed) =
                                diff::compute_diff(&previous_rows, &current_rows);

                            if !added.is_empty() {
                                let res = Self::build_result(
                                    &query.name,
                                    &agent_uuid,
                                    added,
                                    ResultAction::Added,
                                );
                                if tx.send(res).await.is_err() {
                                    tracing::info!(
                                        "[{}] Result channel closed, stopping task.",
                                        query.name
                                    );
                                    break;
                                }
                            }
                            if !removed.is_empty() {
                                let res = Self::build_result(
                                    &query.name,
                                    &agent_uuid,
                                    removed,
                                    ResultAction::Removed,
                                );
                                if tx.send(res).await.is_err() {
                                    tracing::info!(
                                        "[{}] Result channel closed, stopping task.",
                                        query.name
                                    );
                                    break;
                                }
                            }
                        }
                        previous_rows = current_rows;
                    }
                }
            });
        }
    }

    fn build_result(
        query_name: &str,
        agent_uuid: &str,
        rows: Vec<OsqueryRow>,
        action: ResultAction,
    ) -> OsqueryResult {
        let mut result_rows = Vec::with_capacity(rows.len());
        for row in rows {
            let mut columns = Vec::with_capacity(row.len());
            for (k, v) in row {
                columns.push(ColumnEntry { name: k, value: v });
            }
            result_rows.push(OsqueryResultRow { columns });
        }

        OsqueryResult {
            query_name: query_name.to_string(),
            agent_uuid: agent_uuid.to_string(),
            timestamp_ns: Utc::now().timestamp_nanos_opt().unwrap_or(0),
            rows: result_rows,
            action,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_upsert_and_load_queries() {
        let dir = tempfile::tempdir().unwrap();
        let db_path = dir.path().join("queries.db");
        let mut scheduler = QueryScheduler::new(&db_path).unwrap();

        let q1 = ScheduledQuery {
            name: "test_query".to_string(),
            query: "SELECT 1".to_string(),
            interval_secs: 60,
            snapshot: false,
        };

        scheduler.upsert_queries(std::slice::from_ref(&q1)).unwrap();

        let loaded = scheduler.load_queries().unwrap();
        assert_eq!(loaded.len(), 4);
        let q_loaded = loaded.iter().find(|q| q.name == "test_query").unwrap();
        assert_eq!(q_loaded.name, "test_query");
        assert_eq!(q_loaded.query, "SELECT 1");
        assert_eq!(q_loaded.interval_secs, 60);
        assert!(!q_loaded.snapshot);

        let q2 = ScheduledQuery {
            name: "test_query".to_string(),
            query: "SELECT 2".to_string(),
            interval_secs: 120,
            snapshot: true,
        };
        scheduler.upsert_queries(&[q2]).unwrap();

        let loaded = scheduler.load_queries().unwrap();
        assert_eq!(loaded.len(), 4);
        let q_loaded2 = loaded.iter().find(|q| q.name == "test_query").unwrap();
        assert_eq!(q_loaded2.query, "SELECT 2");
        assert_eq!(q_loaded2.interval_secs, 120);
        assert!(q_loaded2.snapshot);
    }
}

File: agent/crates/osquery-client/src/scheduler.py
import asyncio
import logging
import sqlite3
import time
from pathlib import Path
from typing import Dict, List

from osquery_client import OsqueryClient
from diff import compute_diff
from types import (
    ScheduledQuery,
    OsqueryRow,
    OsqueryResult,
    ResultAction,
)

logger=logging.getLogger(__name__)

class QueryScheduler:

    def __init__(self, db_path: str):
        
        self.db_path = Path(db_path)

        self.previous_results: Dict[
            str,
            List[OsqueryRow]
        ] = {}

# Load sql queries

    def load_queries(self) -> List[ScheduledQuery]:
    
        connection = sqlite3.connect(self.db_path)
    
        cursor = connection.cursor()
    
        cursor.execute(
            """
            SELECT
                name,
                query,
                interval_secs,
                snapshot
            FROM scheduled_queries
            """
        )
    
        rows = cursor.fetchall()
    
        connection.close()
    
        queries = []
    
        for row in rows:
    
            queries.append(
    
                ScheduledQuery(
                    name=row[0],
                    query=row[1],
                    interval_secs=row[2],
                    snapshot=bool(row[3]),
                )
    
            )
    
        return queries
    
    #Update or Insert queries
    def upsert_queries(
        self,
        queries: List[ScheduledQuery],
    ):
    
        connection = sqlite3.connect(self.db_path)
    
        cursor = connection.cursor()
    
        for query in queries:
    
            cursor.execute(
                """
                INSERT INTO scheduled_queries
                (name, query, interval_secs, snapshot)
    
                VALUES (?, ?, ?, ?)
    
                ON CONFLICT(name)
    
                DO UPDATE SET
    
                query=excluded.query,
    
                interval_secs=excluded.interval_secs,
    
                snapshot=excluded.snapshot
                """,
    
                (
                    query.name,
                    query.query,
                    query.interval_secs,
                    int(query.snapshot),
                ),
    
            )
    
        connection.commit()
    
        connection.close()
    
    async def run(
        self,
        result_queue: asyncio.Queue,
        socket_path: str,
        agent_uuid: str,
    ):
        """
        Start all scheduled queries.
        """
    
        logger.info("Starting scheduler")
    
        queries = self.load_queries()
    
        tasks = []
    
        for query in queries:
    
            tasks.append(
    
                asyncio.create_task(
    
                    self.query_worker(
                        query,
                        result_queue,
                        socket_path,
                        agent_uuid,
                    )
    
                )
    
            )
    
        await asyncio.gather(*tasks)
    
    
    async def run_query(
        self,
        scheduled_query: ScheduledQuery,
        result_queue: asyncio.Queue,
        socket_path: str,
        agent_uuid: str,
    ):
    
        logger.info(
            "Starting scheduled query %s",
            scheduled_query.name,
        )
    
        client = await OsqueryClient.connect(
            socket_path
        )
    
            try:
    
                response = await client.query(
                    scheduled_query.query
                )
    
                if response.status.code != 0:
      
                    logger.error(
                        "Query %s failed: %s",
                        scheduled_query.name,
                        response.status.message,
                    )
                    return
              
                current_rows = response.rows
              
                previous_rows = self.previous_results.get(
                    scheduled_query.name,
                    [],
                )
              
              #snapshot or differential
                if scheduled_query.snapshot:
                    result = OsqueryResult(
                      name=scheduled_query.name,
                      action=ResultAction.SNAPSHOT,
                      rows=current_rows,
                    )
              
                    await result_queue.put(result)
              
                else:
              
                    added, removed = compute_diff(
                        previous_rows,
                        current_rows,
                    )  
                    if added:
                        result = OsqueryResult(
                          name=scheduled_query.name,
                          action=ResultAction.ADDED,
                          rows=added,
                        )
                        await result_queue.put(result)
                  
                    if removed:
                      result = OsqueryResult(
                          name=scheduled_query.name,
                          action=ResultAction.REMOVED,
                          rows=removed,
                      )    
                      await result_queue.put(result)
                      
              
              #save current rows
              self.previous_results[scheduled_query.name] = current_rows
    
    
            except Exception:
    
                logger.exception(
                    "Query %s failed",
                    scheduled_query.name,
                )
        
        
    
    async def query_worker(self,
                           scheduled_query: ScheduledQuery,
                           result_queue: asyncio.Queue,
                           socket_path: str,
                           agent_uuid: str,
                          ):
        
        while True:
            await self.run_query(
                scheduled_query,
                result_queue,
                socket_path,
                agent_uuid,
        )
        await asyncio.sleep(scheduled_query.interval)
